README.md
Lessn More 2.5.1
================

Homepage: <https://lessnmore.net>   
Source/Fork: <https://github.com/alanhogan/lessnmore>

Lessn More is a free, open-source personal URL shortener with a lot of enhancements over the original 2009 Lessn.

## Features

*	The option to specify custom short URLs (`/rickroll`) when shrinking
*	Your choice of bit.ly-style random URLs (`/fN6rBX`) or incremental (`/9`, `/a`, `/b`…)
*	Bookmarklets for shrinking (and/or tweeting) URLs — with support for custom short URLs
*	[Statistics][statsimg] on redirection usage
*	An [API][] that supports the same commands as the web interface
*	Your choice of all-lowercase or mixed-case short URLs, and whether or not to avoid lookalike characters
*	An optional "banned word list" to prevent auto-generating offensive URLs
*	Theoretically works with a huge number of URLs in the database
*	The ability to add multiple redirections that point to the same long URL
*	Trims punctuation from the right of the slug (the unique part of the short URL), per [best practices][bestp]
*	Compatible with MySQL for sure and _should_ work with Postgres and SQLite (these DBs are not officially supported, so use at your own risk and please contribute fixes)
*	Will return `http:` or `https:` short URLs to match the current protocol your server
	is responding to, by default; you can manually specify your preferred protocol

### Attention to detail

*	Adding a new slug for a URL already in the database will become the "canonical"
	short URL, and will be returned if you ask Lessn More (either by API or not)
	for a short URL to the original resource
*	Lessn More lets you change the character set you want to use
	to generate short URLs on-the-fly with minimal or no "wasted" or
	skipped possible slugs, and yet the insertion algorithm
	is fast. (Not as fast as Lessn 1.0, by necessity, since Lessn 1.0 did not allow
	custom short URLs; but the worst-case insertion time after
	upgrade or a switch of insertion algorithms is on the order of O(log(n)) where
	<var>n</var> is the number of redirections in your database, and the common case is
	on the order of O(1) (constant time).)
*	Compliant with [URL shortener best practices and standards][bestp]
	whenever possible
*	An [easy migration script](https://lessnmore.net/images/installation-script-screen-shot.png) will upgrade your database
	from an existing Lessn or Lessn More installation.

### Caveats

*	It is not recommended to use this shortener in incremental mode
	when there is a good chance that two or more URLs
	will be shrunk at the same time. (Simultaneous reads are, of course, fine.)
*	Lessn More is developed against MySQL and other databases are theoretically supported
	via PDO, but YMMV. Please [report any issues][issues].
*	Changing the character set used for URL generation, when in conjunction with incremental mode,
	is a supported use case but should be done seldom, and with deliberation.

[markdn]:  http://bit.ly/mkdnsyntax   "This document is written in Markdown."
[convert]: http://tinyurl.com/mkdnwmd "Markdown editor with instant HTML preview"

[bestp]:   https://alanhogan.com/tips/rel-shortlink-for-short-urls "Everything you need to know about rel-shortlink and short URLs"

[issues]:  https://github.com/alanhogan/lessnmore/issues "Bugs & Issues on GitHub"
[API]:     https://lessnmore.net/api "Lessn More API documentation"
[statsimg]: https://lessnmore.net/images/stats-this-week-lessn-more.png

Requirements
-------------

* PHP 5.6 or newer
* PHP's PDO (usually included in PHP builds)
* Database: MySQL (most solidly supported), PostgreSQL, or SQLite <small>(though [see this](https://github.com/alanhogan/lessnmore/issues/#issue/7) for SQLite support)</small>
* While your server doesn’t need to be Apache, the project ships with an Apache `.htaccess` file to handle redirecting requests to the Lessn More redirection engine, for your convenience. Michael McKelvaney has written about [using Nginx instead][with-nginx]. For IIS, [bigworm reports success with this](https://github.com/alanhogan/lessnmore/issues/24).

[with-nginx]: http://mckelvaney.co.uk/blog/2013/04/30/deploying-lessn-more-with-nginx-and-try-files/


History
-------

### v1.0

Lessn was the original personal URL shortening service,
written by [Shaun Inman](http://shauninman.com/). It required PHP, MySQL, and mod_rewrite.

### v1.1

Buttered URLs is a Lessn [fork](https://github.com/jfro/butteredurls) by [Jeremy Knope](http://buttered-cat.com/).
Buttered URLs added logging, custom URLs, a migration mechanism, and support for more database types.

### v2.0

Lessn More is a Buttered URLs [fork](https://github.com/alanhogan/lessnmore) by [Alan Hogan](https://alanhogan.com/).
Lessn More increased the robustness of the insertion algorithm,
prevented slug conflicts, updated the bookmarklets, added multiple auto-shorten modes,
banned word lists, and enhanced security.

### v2.1

Trims punctuation from the right of the slug, per [best practices](https://alanhogan.com/tips/rel-shortlink-for-short-urls)

### v2.2

Better stats page.

More compatible with sqlite (see [#7][]).

Other small fixes and improvements.

Supports a custom timezone in `config.php`.

Special thanks to [Matt Wiebe](https://github.com/mattwiebe) for a lot of these changes.

### v2.3

Returns `https://` short URLs when running on a secure server.

Allows shrinking of non-Web URIs such as `ftp:` or `magnet:`.

### 2.4.0

Users logging in over HTTPS will be authenticated with a cookie that will not leak out over HTTPS

Cookies are now set with HTTP_ONLY, which means rogue JavaScript cannot steal your authentication cookie

There is now a LOG OUT button

Lessn More now checks for an updated version of itself whenever you use the web interface. (You still have to manually upgrade!)

### 2.4.1

Stopped checking for new versions of Lessn More whiled logged out, because it will always fail then.

### 2.5.0

**Big news!** Now you can configure Lessn More to generate random slugs of a desired length instead of incremental, guessable slugs. This means you will generate URLs like `/8RTvaN`, `/Nvb3QP` instead of `/a`, `/b`, `/c`, if desired.

The link in LM's header now points to your main admin page,  and the update check info in the footer will always link to the LM project home page.

### 2.5.1

Fixes JavaScript and the Back button on the stats page when on HTTPS.


Legal
-----

Lessn is offered as-is, sans support and without warranty.
Copyright © 2009-10 Shaun Inman and contributors.
Offered under a BSD-like license; see license.txt.

Installation
------------

Installation instructions are different depending on if you are upgrading or doing a fresh install.

All the following refer to files inside the `dist` folder.

### Fresh Install ###

**ONLY** follow these instructions if you are not upgrading!

0. Copy or rename /-/config-example.php to /-/config.php.

1. Open /-/config.php in a plaintext editor and
	create a Lessn username and password then enter your database connection details. You may also choose other settings such as authentication salts, a default home page, and your current time zone.

2. For the shortest URLs possible, upload the contents of this directory (`dist`) to your domain's root public folder.

3. Visit http://doma.in/install.php to create the necessary database tables. (Watch for errors.)

4. Visit http://doma.in/-/ to log in & start using Lessn More! Be sure to grab the bookmarklets.

**NOTE:** If your Lessn'd urls aren't working you probably didn't
upload the .htaccess file. Enable "Show invisible files"
in your FTP application. It's also possible that your host doesn't like
the `<IfModule>` directives; try removed them and just leaving the
`Rewrite*` lines that were wrapped by the `<IfModule>`.
(This seems to happen on 1and1).
Or, maybe you aren’t on Apache. That’s OK, probably — see the Requirements section.

Upgrading
----------

If you are upgrading from a previous version of Lessn or ButteredURLs:

### Upgrading from Lessn 1.0.0 or 1.0.1

1. Using a tool like PhpMyAdmin or the MySQL CLI change the
   checksum index type to INDEX (from UNIQUE).
2.	Continue below with "ALL VERSIONS"

### ALL VERSIONS: Upgrading to Lessn More 2

1.	You are strongly encouraged to back up your database.
1.	Note some old redirections so you can manually check they still work after upgrading (they should, but hey, it's important).
1.	Manually merge your old configs into the new config file.
	There will be new options you will want to make
	decisions about.
1.	Upload all lessn/BU files, excluding config.php, or making sure to use the new one.
1.	Go to http://doma.in/install.php?start=N where
	N is 2 if upgrading from Lessn 1.0, or    
	N is 4 if upgrading from ButteredURLs 1.1.
1.	Test some old known working redirections
1.	Delete install.php.
1.	Grab the new bookmarklets with custom short URL support!

**Congratulations.** You are running the latest version of Lessn More.

### Upgrading from Lessn More 2.0 or 2.1 to LM 2.2

Just upload the latest copy of Lessn More to your server, but skip or delete install.php.

Note that since v2.2, there is only a config-example.php file in the distribution, and no config.php file. You will _not_ need to take care to avoid overwriting your existing config.php file, but do take care not to replace the whole `/-/` folder and your config.php file; please make a backup copy of it before upgrading.

### Upgrading from Lessn More 2.2 to 2.3

You only need the new `/-/index.php` for this upgrade. Optionally, add the new
`PROTOCOL_OVERRIDE` constant from `config-example.php` into your `config.php`.

### Upgrading from 2.3.0 to 2.4.0

Just copy over all the files in `dist/` to your web server’s root (or wherever you previously installed Lessn More). Skip or delete install.php.


### Upgrading from 2.4 to 2.5.0


CHANGES.txt
2.5.1

- Fixes JavaScript and the Back button on the stats page when on HTTPS.


2.5.0

- **Big news!** Now you can configure Lessn More to generate random slugs of a desired length instead of incremental, guessable slugs. This means you will generate URLs like `/8RTvaN`, `/Nvb3QP` instead of `/a`, `/b`, `/c`, if desired.

- The link in LM's header now points to your main admin page,  and the update check info in the footer will always link to the LM project home page.


2.4.1

- Stopped checking for new versions of Lessn More whiled logged out, because it will always fail then.


2.4.0

- Users logging in over HTTPS will be authenticated with a cookie that will not leak out over HTTPS

- Cookies are now set with HTTP_ONLY, which means rogue JavaScript cannot steal your authentication cookie

- There is now a LOG OUT button

- Lessn More now checks for an updated version of itself whenever you use the web interface. (You still have to manually upgrade!)


2.3.0

- Lessn More will now return `http:` or `https:` short URLs to match
	the current protocol your server is responding to, by default.
	You can manually specify your preferred protocol in `config.php`.

- Lessn More lets you shrink non-Web URIs now, too, such as `ftp:` or `magnet:`.

2.2.0

- Better stats page.

- More compatible with sqlite (see [#7][]).

- Other small fixes and improvements.

- Supports a custom timezone in `config.php`.

- Special thanks to [Matt Wiebe](https://github.com/mattwiebe) for a lot of these changes.



2.1.1
- Merged changes from jfro’s fork.
	- Changed insert queries to support Postgres
	- Avoid php short tags which aren't always enabled

2.1.0
- Fixed issue [#4](https://github.com/alanhogan/lessnmore/issues#issue/4) regarding the bookmarklets and clicking ‘Cancel’ in the prompt.
- Fixed issue [#6](https://github.com/alanhogan/lessnmore/issues#issue/6) regarding 404s (thanks to repearhulk for the patch).
- Trims punctuation from the right of the slug, per [best practices](https://alanhogan.com/tips/rel-shortlink-for-short-urls)

2.0.3
- Improved security
- Corrected the stats table

2.0.2
- Optimize a couple queries
- Fix for finding banned words and glyphs anywhere but the end of the slug

2.0.1
- Case sensitivity in MySQL

2.0 (Lessn More fork)
- Protected against SQL injection attacks
- Support for aliases - set the 'url' field in a database row to be the token you want to want to alias to, and the 'redir_type' to 'alias'
- Support for the 410 GONE response for deleted redirections - set 'redir_type' to 'gone' in the DB row
- Avoids collisions between automatically assigned shortcuts and manual ones
- Allows mixed-case shortcuts
- Expanded README
- Added migration script for butteredurls users
- Custom slugs in bookmarklets
- Prettied up installation
- Allowed overwriting existing slugs (with a warning that this is bad for the Internet)

1.1.1
- Fix for #1
- Support for api and redirect URL arguments thanks to Lessn 1.0.5
- Added new configuration option HOMEPAGE_URL for redirecting somewhere if no short url is given

1.1 (butteredurls fork)
- modified query calls to use PDO to allow PostgreSQL/SQLite support
- created an install.php to handle execution of installation queries
- created statistics table to record IP/referer/timestamp
- added column to urls table for doing custom urls like /flickr
- added simple statistics (hits) page in admin area

1.0.2
- updated existing url query to also compare the url in the event of multiple matching checksums
- changed UNIQUE checksum index to non-unique INDEX to allow for legitimate collisions
- added note about .htaccess to README.txt
- added "Updating from 1.0.1" to README.txt
- added CHANGES.txt

1.0.1
- fixed YEAR definition (was 356 days instead of 365)
- fixed typos in README.txt

1.0
- initial release


-/index.php
<?php

/**
*
*	This file allows inserting redirections and displaying stats.
*
*	For redirections added, there are several possibilities:
*
*	1. Adding a new URL, no custom slug: Inserted, and new slug auto-generated.
*
*	2. Adding an existing URL, no custom slug: The old slug is returned.
*
*	3. Adding a new URL, with a custom slug:
*
*	4. Adding an existing URL, with a custom slug: The new slug is inserted anyway.
*		The old redirection(s) become a "aliases" of the new one,
*		so in case 2, the newest slug is returned (it’s "preferred".)
*
*	DEVELOPERS:
*	Note the following possible redir_type values:
*	-	'auto' - Automatically assigned slug. 301 redirect on access.
*	-	'custom' - Manually set slug. 301 redirect on access.
*	-	'alias' - Its 'url' is really just another slug. Do a recursive lookup to redirect on access.
*	-	'gone' - Access results in a 410; should never change
*/
REQUIRE 'config.php';
REQUIRE 'db.php';
REQUIRE 'stats.php';

define('BCURLS_VERSION',	'2.3.0');

define('BCURLS_DOMAIN', 	preg_replace('#^www\.#', '', $_SERVER['SERVER_NAME']));

// Defaults to `http` or `https` per the current request.
// You can hardcode a a protocol in config.php.
if (defined('PROTOCOL_OVERRIDE') && PROTOCOL_OVERRIDE) {
	define('BCURLS_PROTOCOL', PROTOCOL_OVERRIDE);
} else {
	define('BCURLS_PROTOCOL', 'http' . ($_SERVER['HTTPS'] ? 's' : ''));
}

define('BCURLS_URL', 	str_replace('-/index.php', '', BCURLS_PROTOCOL.'://'.BCURLS_DOMAIN.$_SERVER['PHP_SELF']));
define('BCURLS_PATH', 	realpath('.') );

//don't reveal db prefix over HTTP. 16 chars is more than enough to avoid collisions
define('COOKIE_NAME', 	substr(md5(DB_PREFIX.COOKIE_SALT), 4, 16).'auth');
define('COOKIE_VALUE',	md5(USERNAME.PASSWORD.COOKIE_SALT));
define('COOKIE_DOMAIN', '.'.BCURLS_DOMAIN);

if (!defined('API_SALT')) define('API_SALT', 'B75jk4K25M5U7hTAP1'); // added in lessn 1.0.5
define('API_KEY', md5(USERNAME.PASSWORD.API_SALT));

define('NOW', 		time());
define('YEAR',		365 * 24 * 60 * 60);

/**
* @Returns int Characters from the right, 0-(n-1) where n = strlen($slug),
* where a banned word ends. e.g.:
* slug = 'dLpooz'
* 'poo' is a banned word
* returns 1
*/
function bcurls_find_banned_word($slug) {
	global $bcurls_banned_words;
	REQUIRE_ONCE 'banned_words.php';
	foreach ($bcurls_banned_words as $banned){
		$strpos = stripos($slug, $banned);
		if ($strpos !== false) {
			bc_log('Found banned word '.$banned.' in '.$slug);
			// $slug = xBANNEDxx (length 9)
			// $banned = BANNED (length 6)
			// strpos = 1
			// we want to return 2 (BANNED is two characters from the right)
			return strlen($slug) - strlen($banned) - $strpos;
		}
	}
	return FALSE;
}

/**
* @Returns int Characters from the right, 0-(n-1) where n = strlen($slug),
* where a banned glyph ends. e.g.:
*/
function bcurls_find_banned_glyph($slug) {
	if (ADDITIONAL_HOMOGLYPHS_TO_AVOID === false)
		return false;
	$glyphs = str_split(ADDITIONAL_HOMOGLYPHS_TO_AVOID);
	foreach ($glyphs as $banned){
		$strpos = strpos($slug, $banned);
		if ($strpos !== FALSE) {
			bc_log('Found banned glyph '.$banned.' in '.$slug);
			// $slug = xIxx (length 4)
			// $banned = I (length I)
			// strpos = 1
			// we want to return 2 (I is two characters from the right)
			return (strlen($slug) - 1 - $strpos);
		}
	}
	return FALSE;
}

// handle login
if (isset($_POST['username']))
{
	if (md5($_POST['username'].$_POST['password'].COOKIE_SALT) == COOKIE_VALUE)
	{
		setcookie(COOKIE_NAME,
			COOKIE_VALUE,
			NOW + YEAR,
			'/',
			COOKIE_DOMAIN,
			$_SERVER['HTTPS'], // If we are on HTTPS, do not leak cookie to HTTP
			true // HTTPONLY (don't let JS examine the cookie)
		);
		$_COOKIE[COOKIE_NAME] = COOKIE_VALUE;
	}
}
// API login
else if (isset($_GET['api']) && $_GET['api'] == API_KEY)
{
	$_COOKIE[COOKIE_NAME] = COOKIE_VALUE;
}
else if (isset($_GET['api'])) // spit out a nicer failure for API attempts
{
	exit('Invalid API key');
}

// handle logout
if (isset($_GET['logout']))
{
	setcookie(COOKIE_NAME, '', NOW - YEAR, '/', COOKIE_DOMAIN);
	unset($_COOKIE[COOKIE_NAME]);
	header('Location:./');
}

function user_is_logged_in() {
	return isset($_COOKIE[COOKIE_NAME]) && $_COOKIE[COOKIE_NAME] === COOKIE_VALUE;
}

// require login
if (!user_is_logged_in())
{
	include('pages/login.php');
	exit();
}
// prolong login for another year, unless this is an API request
else if (!isset($_GET['api']))
{
	setcookie(COOKIE_NAME,
		COOKIE_VALUE,
		NOW + YEAR,
		'/',
		COOKIE_DOMAIN,
		$_SERVER['HTTPS'], // If we are on HTTPS, do not leak cookie to HTTP
		true // HTTPONLY (don't let JS examine the cookie)
	);
}

// Successfully logged in, so
define('OKAY_TO_SHOW_PAGES', true); //value doesn't matter

/**
*	WARNING! Provides NO checking and should ONLY be used internally!
*	Returns TRUE *or* a string "error code: error info".
*/
function bcurls_insert_url ($url, $checksum, $slug, $redir_type)
{
	global $db;
	$stmt = $db->prepare('INSERT INTO '.DB_PREFIX.'urls (url, checksum, custom_url, redir_type) VALUES(?, ?, ?, ?)');
	$stmt->bindValue(1, $url);
	$stmt->bindValue(2, $checksum);
	$stmt->bindValue(3, $slug);
	$stmt->bindValue(4, $redir_type);
	if($stmt->execute())
		return true;
	else
		return $stmt->errorCode().': '.$stmt->errorInfo();
}
/**
*	WARNING! Provides NO checking and should ONLY be used internally!
*	Returns TRUE *or* a string "error code: error info".
*/
function bcurls_update_slug ($url, $checksum, $slug, $redir_type)
{
	global $db;
	$stmt = $db->prepare('UPDATE '.DB_PREFIX.'urls SET url = ?, checksum = ?, redir_type = ? WHERE custom_url = ? ');
	$stmt->bindValue(1, $url);
	$stmt->bindValue(2, $checksum);
	$stmt->bindValue(3, $redir_type);
	$stmt->bindValue(4, $slug);
	if($stmt->execute())
		return true;
	else
		return $stmt->errorCode().': '.$stmt->errorInfo();
}

/**
* For debuggers, developers, and the curious
*/
function bc_log($message){
	global $bc_log;
	if(!is_string($bc_log)) $bc_log = '';
	if(LOG_MODE) $bc_log .= date('r - ')."$message \n";
}

// new shortcut
if (isset($_GET['url']) && !empty($_GET['url']))
{
	$url = $_GET['url'];
	$slug = NULL;
	$prefix = DB_PREFIX;
	if (!preg_match('#^[^:]+:#', $url))
	{
		$url = 'http://'.$url;
	}
	if(strpos($url, BCURLS_URL) === 0)
	{
		$error = 'You tried to shorten a URL on this domain which should already be short!';
		include('pages/error.php');
		exit;
	}

	// Is there already a row in the DB going to this same URL?
	$checksum 		= (int) sprintf('%u', crc32($url));
	$result = $db->prepare("SELECT id, custom_url, redir_type FROM {$prefix}urls WHERE checksum=? AND "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    .' url = '
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." ? AND url = ? AND redir_type <> 'gone' ORDER BY redir_type DESC LIMIT 1"); //sort so custom is before auto.
	$result->bindValue(1, (int)$checksum);
	$result->bindValue(2, $url);
	$result->bindValue(3, $url);
	if ( ! $result->execute())
	{
		$error = 'Problem executing query to check if the URL is already in the DB! '
			.$result->errorCode().': '.$result->errorInfo();
		include('pages/error.php');
		exit;
	}

	if ($row = $result->fetch(PDO::FETCH_ASSOC))
	{
		$existing_slug = $row['custom_url'];
	} else {
		$existing_slug = false;
		if(LOG_MODE) bc_log('No existing slug'); //debug
	}

	if(isset($_GET['custom_url']) && strlen(trim($_GET['custom_url'])))
	{	// user wants to assign a custom short URL
		$custom_url = trim($_GET['custom_url']);
		// check if the slug is already in use
		$stmt = $db->prepare("SELECT * FROM {$prefix}urls WHERE "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    .' custom_url = '
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    .' ? AND  custom_url =  ?');
		$stmt->bindValue(1, $custom_url);
		$stmt->bindValue(2, $custom_url);
		$stmt->execute();
		$row = $stmt->fetch(PDO::FETCH_ASSOC);
		if($row && ! isset($_GET['overwrite']))
		{
			$error = 'The custom short URL you attempted to use (/'.$row['custom_url'].') is already in use, and is '.($row['redir_type'] == 'alias' ? 'an alias for /': 'pointing to ').$row['url'];
			if(isset($_GET['api'])) exit('Error: Token in use');
			else
			{
				include('pages/add.php');
				exit;
			}
		}
		elseif ($row) //implicit: they allowed overwrite
		{

			$redir_type = 'custom';
			$slug = $custom_url;
			// Update
			$insert_result = bcurls_update_slug ($url, $checksum, $slug, $redir_type);
			if($insert_result !== true) {
				$error = $insert_result;
				include('pages/error.php');
				exit;
			}
		}
		else
		{
			$redir_type = 'custom';
			$slug = $custom_url;
			// Insert!
			$insert_result = bcurls_insert_url ($url, $checksum, $slug, $redir_type);
			if($insert_result !== true) {
				$error = $insert_result;
				include('pages/error.php');
				exit;
			}

		}

		// Old records become aliases to this one
		if($existing_slug !== false)
		{
			// User added a new custom short URL even though that URL is already in the DB
			// Update old redirections so they are no more than aliases of the new one ;)
			$update_to_alias_sql = "UPDATE {$prefix}urls SET redir_type = 'alias', url = :slug, checksum = :newchecksum WHERE checksum = :checksum AND url = :url AND "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." url = "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." :url AND (redir_type = 'custom' OR redir_type = 'auto') AND "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." custom_url <> "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." :slug";
			$updt_a = $db->prepare($update_to_alias_sql);
			$updt_a->execute(array(
				'checksum'		=> $checksum,
			 	'url'			=> $url,
				'slug'			=> $slug,
				'newchecksum' 	=> (int) sprintf('%u', crc32($slug)),
			));
			unset($updt_a);
		}
	}
	elseif($existing_slug == false)
	{	// auto-assign a slug for this new URL.
		$redir_type = 'auto';
		require_once 'library/BaseIntEncoder.php';

		// PREPARE
		switch(AUTO_SLUG_METHOD) {
			case 'base62':
				$glyphs = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
				$base = 62;
				break;
			case 'mixed-smart':
				$glyphs = '23456789abcdefghijkmnpqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ';
				// we excluded 6 characters: 0oO1Il
				$base = 56;
				break;
			case 'base36':
				$glyphs = '0123456789abcdefghijklmnopqrstuvwxyz';
				$base = 36;
				break;
			case 'smart':
				$glyphs = '23456789abcdefghijkmnpqrstuvwxyz';
				$base = 32; //exclude 0o1l
				break;
			default:
				throw new Exception ('Unsupported method to generate unique slugs!');
				break;
		}
		$seek_count = 0; // When we need to skip tweets
		$auto_seek_pow = 1.1513;
		$custom_seek_pow = 1.031;
		$auto_assign_sql = "SELECT base10 FROM {$prefix}autoslug "
			.'WHERE method = :method LIMIT 1';
		$auto = $db->prepare($auto_assign_sql);
		$auto->execute(array('method'=>AUTO_SLUG_METHOD));
		$counter = $auto->fetch(PDO::FETCH_ASSOC);
		$counter = $counter['base10'];

		// For binary search //TODO
		$high = $low = $counter;

		$total_attempts_remaining = 250;

		// Check to insert (seek)
		while ($slug === NULL) {
			// Never just try forever
			if($total_attempts_remaining-- < 1) {
				bc_log('Total attempts maxed out in insert loop');
				$error = 'Tried too many times seeking an available slug';
				include('pages/error.php');
				exit;
			}

			$slug = BaseIntEncoder::encode($counter, $glyphs, $base);

			// Check if slug is free
			$stmt = $db->prepare("SELECT custom_url, redir_type FROM {$prefix}urls WHERE "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." custom_url = "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." :slug AND custom_url = :slug");
			$stmt->execute(array('slug'=>$slug));
			$row = $stmt->fetch(PDO::FETCH_ASSOC);
			if( ! $row ) //okay to insert
			{
				$high = $counter;
				break; // found a suitable URL
			}
			else
			{
				$low = $counter;
				$counter += ceil(pow(++$seek_count,
					($row['redir_type'] == 'custom'
						? $custom_seek_pow
						: $auto_seek_pow
					)
				));
				if(LOG_MODE) bc_log('Slug '.$slug.' already in use; its redir_type is '.$row['redir_type']
					." - ".'Incremented counter from '.$low.' to '.$counter);
				$slug = NULL;
				continue;
			}

		}


		// Binary Search
		if(LOG_MODE) bc_log("Before binary search: low: $low; high: $high; slug: $slug");
		// Note: Low is always "known bad" and high is always "known good"
		$high = (string)$high;
		$low = (string)$low;
		while($low != $high)
		{
			if($high == bcadd((string)$low,'1'))
			{
				$counter = $high;
				$slug = BaseIntEncoder::encode($counter, $glyphs, $base);
				if(LOG_MODE) bc_log('Binary search decided to use '.$slug.' (counter '.$counter.") because high == low+1");
				break;
			}

			$counter = bcadd($low, bcmul(bcsub($high, $low), '0.5', 0)); // at least +1
			$slug = BaseIntEncoder::encode($counter, $glyphs, $base);
			$stmt = $db->prepare("SELECT custom_url, redir_type FROM {$prefix}urls WHERE  custom_url = :slug AND "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." custom_url = "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." :slug");
			$stmt->execute(array('slug'=>$slug));
			$row = $stmt->fetch(PDO::FETCH_ASSOC);
			if( ! $row ) // empty spot in the DB!
			{
				$high = $counter;
				if(LOG_MODE) bc_log($slug.' was available (counter: '.$counter.')');

			}
			else
			{
				$low = $counter;
				if(LOG_MODE) bc_log($slug.' was occupied (counter: '.$counter.')');
			}

		}


		$total_attempts_remaining += 50;
		$validated = true;
		// (Carefully, loopingly) Insert!
		while ($slug !== false){
			// Never just try forever
			if($total_attempts_remaining-- < 1) {
				bc_log('Total attempts maxed out in insert loop');
				$error = 'Tried too many times';
				include('pages/error.php');
				exit;
			}

			if( ! $validated){
				$stmt = $db->prepare("SELECT custom_url, redir_type FROM {$prefix}urls WHERE  custom_url = :slug AND "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." custom_url = "
    .(DB_DRIVER === 'sqlite' ? '' : 'BINARY')
    ." :slug");
				$stmt->execute(array('slug'=>$slug));
				$row = $stmt->fetch(PDO::FETCH_ASSOC);
				if($row)
				{
					$counter = bcadd($counter, '1');
					$slug = BaseIntEncoder::encode($counter, $glyphs, $base);
					continue;
				}
			}

			if(USE_BANNED_WORD_LIST){
				$banned_pos = bcurls_find_banned_word($slug);
				if($banned_pos !== FALSE) {
					if($banned_pos > 0) {
						// If slug is e.g. BANNEDa
						// we want to increment counter by (10 - a) in base $base
						// to make the slug BANNEE0
						$rollover = bcpow((string)$base, (string)$banned_pos);
						bc_log('Rollover calculated: '.$rollover);
						// and computure that a(base $base) = 11(base 10)
						$already_in = BaseIntEncoder::decode(substr($slug, 0-$banned_pos),
							$glyphs, $base);
						bc_log('Already_in calculated: '.$already_in.' based on "'.substr($slug, 0-$banned_pos).'"');
						// so (10 in base $base) in base 10 - 11 in base 10
						$diff = bcsub($rollover,$already_in);
						$counter = bcadd($counter, $diff);
						if(LOG_MODE) bc_log('Counter += '.$diff.
							' for banned word,  is now '.$counter.' - slug '.$slug);
					} else {
						$counter = bcadd($counter, '1');
						if(LOG_MODE) bc_log('Counter++ for banned word in slug '.$slug
							.', counter is now '.$counter);
					}
					$slug = BaseIntEncoder::encode($counter, $glyphs, $base);
					$validated = false;
					continue;
				}
			}
			if(ADDITIONAL_HOMOGLYPHS_TO_AVOID){
				$banned_pos = bcurls_find_banned_glyph($slug);
				if($banned_pos !== FALSE) {
					if($banned_pos > 0) {
						$rollover = bcpow((string)$base, (string)$banned_pos);
						bc_log('Rollover calculated: '.$rollover);
						$already_in = BaseIntEncoder::decode(substr($slug, 0-$banned_pos),
							$glyphs, $base);
						bc_log('Already_in calculated: '.$already_in.' based on "'.substr($slug, 0-$banned_pos).'"');
						$diff = bcsub($rollover,$already_in);
						$counter = bcadd($counter, $diff);
						if(LOG_MODE) bc_log('Counter += '.$diff.
							' for homoglyphs,  is now '.$counter.' - slug '.$slug);
					} else {
						$counter = bcadd($counter, '1');
						if(LOG_MODE) bc_log('Counter++ for homoglyph in slug '.$slug
							.', counter is now '.$counter);
					}
					$slug = BaseIntEncoder::encode($counter, $glyphs, $base);
					$validated = false;
					continue;
				}
			}

			// Actually insert
			try{
				$insert_result = bcurls_insert_url ($url, $checksum, $slug, $redir_type);
				if($insert_result !== true) {
					bc_log('Insertion result (not true)'.(string)$insert_result);
					$counter = bcadd($counter, '1');
					$slug = BaseIntEncoder::encode($counter, $glyphs, $base);
					$validated = false;
					continue;
				}
			} catch(Exception $e){
				bc_log('Exception inserting or incrementing counter.'.$e);
				$error = $e;
				include('pages/error.php');
				exit;
			}
			bc_log('About to increment counter');
			// Update counter
			try{
				// Update counter
				// Note it would be great if this was a strictly incremental
				// and atomic operation: "If your current base10 is smaller
				// than mine, change it to mine, database!"
				// This would deal with simultaneous inserts.
				$counter_update_sql = "UPDATE {$prefix}autoslug
					SET base10 = :base10
					WHERE method = :method LIMIT 1";
				$ctr_up = $db->prepare($counter_update_sql);
				$ctr_up_res = $ctr_up->execute(array(
					'method' => AUTO_SLUG_METHOD,
					'base10' => bcadd((string)$counter,'1'
				)));
				if(! $ctr_up_res) {
					$error = 'Could not update the autoslug index (this will hurt insertion performance.) '
						.$ctr_up->errorCode().': '.$ctr_up->errorInfo();
					bc_log('Counter increment error. '.$error);
				}
			}
			catch (Exception $e2) {
				$error = (string) $e2;
				bc_log('Could not update counter. '.$e2);
			}

			break;
		}

		if(LOG_MODE) bc_log('$total_attempts_remaining after finish: '.$total_attempts_remaining);
	}
	else
	{	// This is already in the DB, don't provide a new one
		$slug = $existing_slug;
	}


	$new_url = BCURLS_URL.$slug;

	if (isset($_GET['tweet']))
	{
		$_GET['redirect'] = 'https://twitter.com/?status=%l';
	}
	if (isset($_GET['redirect']))
	{
		header('Location: '.str_replace('%l', urlencode($new_url), $_GET['redirect']));
		exit();
	}
	if (isset($_GET['api']))
	{
		echo $new_url;
		exit();
	}

	include('pages/done.php');
}
else if(isset($_GET['stats']))  // Display stats
{
	$top_urls = stats_top_urls($db);
	$top_referers = stats_top_referers($db);
	$todays_urls = stats_todays_stats($db);
	$weeks_urls = stats_thisweeks_stats($db);
	$number_lessnd = stats_total_lessnd($db);
	$number_redirected = stats_total_redirects($db);
	include('pages/stats.php');
}
elseif(isset($_GET['mark_gone']) && isset($_GET['slug']) && strlen(trim($_GET['slug'])))
{ // Mark a redirection as GONE
	// TODO
}
elseif(isset($_GET['current_release']))
{
	require('current_release.php');
	exit();
}
else
{
	include('pages/add.php');
}
